Update the heuristic for AArch64 bmm/baddbmm (F32 only)

michalowski-arm opened the pull request:
Updates heuristic for bmm/baddbmm and consolidates all heuristic logic in a single location

 - The goal of the consolidation is to improve maintainability and readability of the heuristic logic. Instead of different parts scattered across two files, this patch centralizes everything inside `Matmul.cpp`, where there already exists heuristic-based selection for mkldnn.
 - The logic of the check itself doesn't change (existing code is reused where possible) but a separate heuristic threshold for bmm/baddbmm is introduced based on newer, benchmarking data. Use the script below to see the performance improvement for bmm from the new heuristic:
 ```
import torch
import time

# Set below to True to use cases selected by only one of the hueristics.
USE_ONLY_DIVERGENT_TEST_CASES = True   
BATCH_SIZES  = [ 1, 8, 32, 64, 128, 256 ]
M_DIMS       = [ 4, 8, 16, 32, 64, 256, 512 ]
N_DIMS       = [ 4, 8, 16, 32, 64, 256, 512 ]
K_DIMS       = [ 4, 8, 16, 32, 64, 256, 512 ]
ITERS = 50

def old_heuristic(m, n, k):
     is_above_min_dims = m > 8 and n > 8 and k > 8
     is_above_min_size = m*n*k > 8_192
     return is_above_min_dims and is_above_min_size

def new_heuristic(b, m, n, k):
     return b*b*m*n*k >= 4_194_304

def generate_test_cases():
    test_cases = []
    for b in BATCH_SIZES:
        for m in M_DIMS:
            for n in N_DIMS:
                    for k in K_DIMS:
                        if USE_ONLY_DIVERGENT_TEST_CASES:
                            if old_heuristic(m, n, k) != new_heuristic(b, m, n, k):
                                test_cases.append([b, m, n, k])
                        else:
                            test_cases.append([b, m, n, k])
    return test_cases

def test(x, y):
    for _ in range(5):
        torch.bmm(x, y)
    perf = 0.0
    for _ in range(ITERS):
        start = time.time()
        torch.bmm(x, y)
        end = time.time()
        perf += (end - start) / ITERS
    return perf

def main():
    print(f"{'b':<10}{'m':<10}{'n':<10}{'k':<10}{'time (s)':10}")
    cumulative_mean_time = 0.0
    for b, m, n, k in generate_test_cases():
        mean_time = test(torch.rand(b, m, n), torch.rand(b, n, k))
        cumulative_mean_time += mean_time
        print(f"{b:<10}{m:<10}{n:<10}{k:<10}{mean_time:10.3e}")
    print(f"Cumulative mean time = {cumulative_mean_time:.4f} s")


if __name__ == "__main__":
    main()
```

From the script we see that cumulative mean time from all test cases (at 16 threads) is:
 - 1.6195 s for the old heuristic
 - 0.7012 s for the new heuristic

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @snadampal @milpuz01 @nikhil-arm @fadara01 @robert-hardwick @nWEIdia @malfet

michalowski-arm committed:
Update the heuristic for AArch64 bmm/baddbmm

malfet reviewed (commented):
Can you provide a bit more description on what this consolidation is trying to do?
Or is it just a better engineering change?
If it changes the semantic of the check, would be nice to have some sort of unit test added, if it's a performance only change would be nice to have script in PR description that one can run to observe those perf changes

michalowski-arm commented:
> Can you provide a bit more description on what this consolidation is trying to do? Or is it just a better engineering change? If it changes the semantic of the check, would be nice to have some sort of unit test added, if it's a performance only change would be nice to have script in PR description that one can run to observe those perf changes

Added some extra information in the PR description, let me know if anything more is needed.

aditew01 commented on aten/src/ATen/native/mkldnn/Matmul.cpp:
```diff
@@ -336,10 +372,26 @@ inline bool checksize(const Tensor& mat1, const Tensor& mat2){
+#if defined(__aarch64__)
+    const int64_t mkldnn_acl_bmm_baddbmm_threshold = get_mkldnn_acl_bmm_baddbmm_threshold();
+    // BATCH_SIZE^2 * M * N * K >= THRESHOLD
+    return mat1.size(0) * mat1.size(0) * mat1.size(1) * mat1.size(2) * mat2.size(2) >= mkldnn_acl_bmm_baddbmm_threshold;
```
For Neoverse cores, `mkldnn_acl_bmm_baddbmm_threshold` == `1L<<22`
Does this imply we don't want to dispatch to MKLDNN for  aten::baddbmm or aten::bmm cases?
 
The script attached only tests for float32 cases. Did we test the performance for float16 / bfloat16?

fadara01 replied:
>For Neoverse cores, mkldnn_acl_bmm_baddbmm_threshold == 1L<<22
Does this imply we don't want to dispatch to MKLDNN for aten::baddbmm or aten::bmm cases?

`1L<<22` is `4,194,304` which is not as large as it seems.
for `K, N = 1024` (which is not large by today's standards), we'll got to oneDNN/ACL with `M=4` and `batch size = 1`

malfet commented on aten/src/ATen/native/mkldnn/Matmul.cpp:
```diff
@@ -322,6 +322,42 @@ void mkldnn_matmul(
 
 }
 
+#if defined(__aarch64__)
```
This should be guarded behind USE_MKLDNN_ACL or something like that

fadara01 commented on aten/src/ATen/native/mkldnn/Matmul.cpp:
```diff
@@ -336,10 +372,26 @@ inline bool checksize(const Tensor& mat1, const Tensor& mat2){
+#if defined(__aarch64__)
+    const int64_t mkldnn_acl_bmm_baddbmm_threshold = get_mkldnn_acl_bmm_baddbmm_threshold();
+    // BATCH_SIZE^2 * M * N * K >= THRESHOLD
+    return mat1.size(0) * mat1.size(0) * mat1.size(1) * mat1.size(2) * mat2.size(2) >= mkldnn_acl_bmm_baddbmm_threshold;
```
where's the `BATCH_SIZE^2` in the equation?

fadara01 commented:
@pytorchbot label "ciflow/linux-aarch64"

fadara01 commented on aten/src/ATen/native/mkldnn/Matmul.cpp:
```diff
@@ -336,10 +372,26 @@ inline bool checksize(const Tensor& mat1, const Tensor& mat2){
+#if defined(__aarch64__)
+    const int64_t mkldnn_acl_bmm_baddbmm_threshold = get_mkldnn_acl_bmm_baddbmm_threshold();
+    // BATCH_SIZE^2 * M * N * K >= THRESHOLD
+    return mat1.size(0) * mat1.size(0) * mat1.size(1) * mat1.size(2) * mat2.size(2) >= mkldnn_acl_bmm_baddbmm_threshold;
```
What's the rationale behind squaring the batch size?
Why can't we just use `mat1.size(0) * mat1.size(1) * mat1.size(2) * mat2.size(2)`?
and add a `mkldnn_acl_bmm_min_size` instead of "threshold"

michalowski-arm replied:
The heuristic is derived purely experimentally. I generated large samples of bmm calls which used OpenBLAS only and then an equivalent one which only used Compute Library. I then used linear regression to find out the best heuristic, with the caveat that the regression was done over logarithms of dimensions so that I could see relative contribution of each dimension to the heuristic. This is how batch size got squared -- it really does not need to be squared specifically, the power just needed to be double that of any other dimension. If it wasn't, the best heuristic I could have got was about 70% accurate (i.e. 70% of the calls tested were correctly sent to the backend which was faster in that case). With batch size's power twice that of other dimensions, we get around 80-90% accuracy (varies with number of threads). For reference, the old heuristic had about 50% accuracy in my tests.

fadara01 replied:
Nice! That's for the great explanation :)

michalowski-arm committed:
Replace defined(__aarch64__) guard in Matmul.cpp

Replaced the defined(__aarch64__) with AT_MKLDNN_ACL_ENABLED() in the
parts of Matmul.cpp relevant to the bmm heuristic

fadara01 commented:
@pytorchbot label "ciflow/linux-aarch64"

fadara01 commented:
@malfet @aditew01 can you guys give this another look?

aditew01 reviewed (dismissed):
LGTM!

malfet reviewed (dismissed):
LGTM

malfet commented:
@pytorchbot merge

malfet commented:
FYI this change is being reverted right now as causing 50% perf regression...

jeanschmidt commented:
@pytorchbot revert -m "breaking internal models, @malfet may you help merge this?" -c ghfirst

aditew01 commented:
@malfet @jeanschmidt where can I look for the model break / regression suite to fix the heuristics ?

michalowski-arm commented:
@malfet @jeanschmidt In this CI run we don't see any regressions: [TorchInductor Performance DashBoard](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Fri%2C%2023%20May%202025%2014%3A53%3A56%20GMT&stopTime=Fri%2C%2030%20May%202025%2014%3A53%3A56%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cpu%20(aarch64)&lBranch=test/bmm_heur&lCommit=5b57dbc93fa6187356e7a1eda1f3bbb0a4684035&rBranch=main&rCommit=5623d3022840a71f6417f956b21a5d190c2638ea). This was for bf16 only, I don't think we can run f32. Are there other tests we are supposed to be running? Also, presumably the regression you saw was for AArch64, or did this change somehow impact x86 performance?

malfet commented:
@pytorchbot rebase

aditew01 commented:
@pytorchbot rebase

michalowski-arm committed:
Add FBCODE guard to the hueristic on aarch64

malfet reviewed (changes requested):
Didn't we have a discussion that this heuristic should only apply to bf16? (and that was the reason for revert)

aditew01 commented:
@malfet CI is green.

aditew01 commented:
@pytorchbot rebase
